feat: add prepare release task

jackd248 opened the pull request:
First approach of a structured release preparation task to automate all including steps.

Further work on the following steps:

- npm update
- npm build
- npm qs
- php qs
- acceptance test

Defcon0 reviewed (changes requested):
Looks good so far. Two things I noticed while testing:

1. You use the abbreviation qs (like in php_qs). Shouldn't it be qa (quality assurance) in an English context?
2. I had issues with the pre commit hooks set up in captainhook when running the task in the ddev container since "ddev" is not defined in the container itself:

```
ddev exec vendor/bin/dep develop:release:prepare local --newVersion=3.7.0
...
[local] run git commit -m "build: version 3.7.0-RC"
[local] err .git/hooks/pre-commit: 12: ddev: not found
[local] exit code 1 (General error)
ERROR: Task develop:release:prepare failed!
Failed to execute command vendor/bin/dep develop:release:prepare local --newVersion=3.7.0: exit status 1
```

When running it outside of ddev (`vendor/bin/dep develop:release:prepare local --newVersion=3.7.0`) it worked. But this implies that I need to have dependencies locally installed which isn't a good practice. Maybe we need to cover that in the captainhook.json so that we have a distinction between the "inside-ddev case" and the "outside-ddev-case"?

Defcon0 commented on deployer/develop/task/prepare_release.php:
```diff
@@ -0,0 +1,201 @@
+
+    checkPreconditions();
+
+    info("The following steps will be executed:\n - tabula_rasa \n - start_new_release \n - " . implode("\n - ", get('develop_steps')));
```
Has to be `develop_release_steps` else throws errors.

jackd248 replied:
I removed this steps.

Defcon0 commented on deployer/develop/task/prepare_release.php:
```diff
@@ -0,0 +1,201 @@
+    /* new release version */
+    debug("set new version: \"{$version}-RC\"");
+    runLocally("composer config version {$version}-RC --working-dir " . get('composer_path_app'));
+    runLocally("composer update nothing -q --no-scripts --no-plugins --working-dir " . get('composer_path_app'));
```
This throws errors in drupal context:

```
vendor/bin/dep develop:release:prepare local --newVersion=3.7.0
task develop:release:prepare
[local] info The following steps will be executed:
 - tabula_rasa 
 - start_new_release 
 - composer_update_ci
 - composer_update_app
[local] info Which steps shall be executed?
[local] composer_update_ci (default: yes) 
[local] composer_update_app (default: yes)   
[local] info Start preparation ...
[local] info [step] tabula rasa
[local] info [step] start new release
[local]  error  in prepare_release.php on line 130:
[local] run composer update nothing -q --no-scripts --no-plugins --working-dir ./app
[local] err Your requirements could not be resolved to an installable set of packages.
[local] exit code 2 (Misuse of shell builtins)
```

When removing the `--no-plugins` it works.

Also adding `--no-scripts` causes the robots.txt not to be adjusted resulting in a commit with a plain robots.txt without our adjustments in the commit. Later on this gets fixed when updating the app "normally" (without `--no-scripts`), but it's a bit bad nevertheless.

Maybe we should parameterize this like so:

```
runLocally("composer update nothing -q " . get('composer_app_release_options') . " --working-dir " . get('composer_path_app'));
```

with the default value `--no-scripts --no-plugins` in the `set.php`?

jackd248 replied:
I added this configuration for now, but I'm not so happy with this, because this is just the composer function to update the release number and not all kind of composer release options. Maybe need some better naming or so.

Defcon0 replied:
Yep, the naming was kind of a shot in the dark 😅️ Thanks for adding.

jackd248 commented:
Hej @Defcon0,
I rewrote the tasks to follow more the deployer concept of adjusting tasks. So it should be much easier to configure the release steps to the project needs.

> You use the abbreviation qs (like in php_qs). Shouldn't it be qa (quality assurance) in an English context?

You're absolutely right, I corrected this.

 > I had issues with the pre commit hooks set up in captainhook when running the task in the ddev container since "ddev" is not defined in the container itself

For now I added `--no-verify` to the git commit to skip git hooks. This could be configurable for the future, but the ddev + captainhook thing is another construction site.
